src/main/java/rx/observables/AsyncOnSubscribe.java

```diff
@@ -0,0 +1,510 @@
+                    try {
+                        onNextCalled = false;
+                        nextIteration(r);
+                        if (onNextCalled)
+                            r = poll();
+                        if (hasTerminated || isUnsubscribed()) {
+                            parent.onUnsubscribe(state);
+                        }
+                    } catch (Throwable ex) {
+                        handleThrownError(parent, state, ex);
+                        return;
+                    }
+                } while (r != null && !hasTerminated);
+            }
+        }
+
+        private void handleThrownError(final AsyncOnSubscribe<S, T> p, S st, Throwable ex) {
+            if (hasTerminated) {
+                RxJavaPlugins.getInstance().getErrorHandler().handleError(ex);
+            } else {
+                hasTerminated = true;
+                merger.onError(ex);
+                unsubscribe();
+            }
+        }
+
+        @Override
+        public void onCompleted() {
+            if (hasTerminated) {
+                throw new IllegalStateException("Terminal event already emitted.");
+            }
+            hasTerminated = true;
+            merger.onCompleted();
+        }
+
+        @Override
+        public void onError(Throwable e) {
+            if (hasTerminated) {
+                throw new IllegalStateException("Terminal event already emitted.");
+            }
+            hasTerminated = true;
+            merger.onError(e);
+        }
+
+        // This exists simply to check if the subscription has already been
+        // terminated before getting access to the subscription
+        private static Subscription SUBSCRIPTION_SENTINEL = new BooleanSubscription();
+
+        @Override
+        public void onNext(final Observable<? extends T> t) {
+            if (onNextCalled) {
+                throw new IllegalStateException("onNext called multiple times!");
+            }
+            onNextCalled = true;
+            if (hasTerminated)
+                return;
+            subscribeBufferToObservable(t);
+        }
+
+        private void subscribeBufferToObservable(final Observable<? extends T> t) {
+            BufferUntilSubscriber<T> buffer = BufferUntilSubscriber.<T> create();
+            final AtomicReference<Subscription> holder = new AtomicReference<Subscription>(null);
+            Subscription innerSubscription = t
+                .doOnTerminate(new Action0() {
+                    @Override
+                    public void call() {
+                        if (!holder.compareAndSet(null, SUBSCRIPTION_SENTINEL)) {
+                            Subscription h = holder.get();
+                            subscriptions.remove(h);
+                        }
+                    }})
+                .subscribe(buffer);
+
+            if (holder.compareAndSet(null, innerSubscription)) {
+                subscriptions.add(innerSubscription);
```

Comment:
I'm not sure about the thread-safety of accessing `subscriptions`. If doOnTerminate happen to run on a different thread, `subscriptions` may get corrupted and even if backed by CHM, the remove on line 458 may happen just before the add on line 464 and `subscriptions` now leaks the reference. 

What would be better is to make sure `subscriptions` always holds that needs to be removed by adding the innerSubscription before calling t.subscribe() which also means you need to pre-wrap `buffer` so that subscribe returns the same instance. This way, the wrapped buffer instance is added before the call to subscribe and will be guaranteed to be removed when `doOnTerminate` is executed.